components/plugin/src/miniboxing/plugin/metadata/MiniboxMetadataUtils.scala

```diff
@@ -168,59 +133,29 @@ trait MiniboxMetadataUtils {
-                      if (p.owner.isArray)
-                        useMbArrayInsteadOfArrayWarning(p)
-                      else if (!p.hasAnnotation(SpecializedClass))
-                        suboptimalCodeWarning(pos, s"The ${p.owner.tweakedFullString} would benefit from miniboxing type " +
-                                                   s"parameter ${p.nameString}, since it is instantiated by miniboxed " +
-                                                   s"type parameter ${tpar.nameString.stripSuffix("sp")} of " +
-                                                   s"${metadata.getStem(tpar.owner).tweakedToString}.",
-                                                   p.isGenericAnnotated, inLibrary = !common.isCompiledInCurrentBatch(p))
-                      else
-                        replaceSpecializedWithMiniboxedWarning(p)
+                    (new BackwardWarningForMiniboxedTypeParam(p, tpar, spec, pos, inLibrary = !common.isCompiledInCurrentBatch(p))).warn()
                     (p, spec)
 
-                  case None if metadata.miniboxedTParamFlag(tpar.deSkolemize) && metadata.isClassStem(tpar.deSkolemize.owner) && !p.isMbArrayMethod =>
-                    if (metadata.miniboxedTParamFlag(p) && !isUselessWarning(p.owner))
-                      if (!p.hasAnnotation(SpecializedClass))
-                        suboptimalCodeWarning(pos, "The following code could benefit from miniboxing specialization " +
-                                                   "(the reason was explained before).",
-                                                   p.isGenericAnnotated, inLibrary = !common.isCompiledInCurrentBatch(p))
-                      else
-                        replaceSpecializedWithMiniboxedWarning(p)
-                    (p, Boxed)
-
                   case None =>
-                    if (metadata.miniboxedTParamFlag(p) && !isUselessWarning(p.owner))
-                      if (!tpar.hasAnnotation(SpecializedClass))
-                        suboptimalCodeWarning(pos, s"The following code could benefit from miniboxing specialization " +
-                                                   s"if the type parameter ${tpar.name} of ${tpar.owner.tweakedToString} " +
-                                                   s"""would be marked as "@miniboxed ${tpar.name}" (it would be used to """ +
-                                                   s"instantiate miniboxed type parameter ${p.name} of ${p.owner.tweakedToString})",
-                                                   p.isGenericAnnotated)
-                      else
-                        replaceSpecializedWithMiniboxedWarning(tpar)
+                    (new BackwardWarningForStemClass(p, tpar, pos, inLibrary = !common.isCompiledInCurrentBatch(p))).warn()
+                    (new BackwardWarningForInnerClass(p, tpar, pos, inLibrary = !common.isCompiledInCurrentBatch(p))).warn()
```

Comment:
I understand the logic, but it's a bit counter-intuitive: you create both warnings, but you're sure only one will trigger. Why not move everything in two possible warnings: `ForwardWarning` and `BackwardWarning` -- and add the logic to decide what to do in those two classes? It's a more complex refactoring, but it also moves all the logic and special-casing there. IMO, the signatures for the classes should be:

```
case class ForwardWarning(pos: Position, mboxedTypeParam: Symbol, nonMboxedType: Type)
case class BackwardWarning(pos: Position, nonMboxedTypeParam: Symbol, mboxedType: Type)
```

And when you write:

```
if (metadata.miniboxedTParamFlag(p))
  BackwardWarning(pos, nonMboxedTypeParam = p, mboxedType = tpe)
```
You immediately realize it's incorrect: actually, `p` is miniboxed and `tpe` is not, so it's a forward warning :)